modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheAdapter.java

```diff
@@ -3653,6 +3654,52 @@ public void awaitLastFut() {
         }
     }
 
+    /**
+     * Replaces previous async operation future on transaction suspend.
+     */
+    public @Nullable FutureHolder suspendLastFut() {
+        FutureHolder holder = lastFut.get();
+
+        IgniteInternalFuture fut = holder.future();
+
+        if (fut != null && !fut.isDone()) {
+            lastFut.set(new FutureHolder());
+
+            return holder;
+        }
+        else
+            return null;
+    }
+
+    /**
+     * Replaces previous async operation future on transaction resume.
+     */
+    public void resumeLastFut(FutureHolder holder) {
+        IgniteInternalFuture resumedFut = holder.future();
+
+        if (resumedFut == null || resumedFut.isDone())
+            return;
+
+        FutureHolder threadHolder = lastFut.get();
+
+        IgniteInternalFuture threadFut = threadHolder.future();
+
+        if (threadFut != null && !threadFut.isDone()) {
+            threadHolder.lock();
+
+            try {
+                GridCompoundFuture f = new GridCompoundFuture<>().add(threadFut).add(resumedFut).markInitialized();
```

Comment:
I'm not sure, but shouldn't we use here `EmbeddedFuture` instead of `CompoundFuture`? Like it is done in `GridCacheAdapter#asyncOp(GridNearTxLocal,GridCacheAdapter.AsyncOp<T>, CacheOperationContext, boolean)`

Reply:
EmbeddedFuture executes asynchroniously some code after completion of one future. Here we need to wait for two futures (threadFut and resumedFut) and don't have any additional code to execute async (except code in saveFuture), so I have no idea how EmbeddedFuture will help here.